### PR TITLE
Implement UniTensor arange/linspace with shape

### LINKED ISSUE #325: UniTensor.arange(int lst) does not exist
It would be great if arange would work with different shapes then vectors as well.
In the draft, there is a codeline:
```
uT = cytnx.UniTensor.arange([2,3,4], labels=["a","b","c"], name="tensor uT")
```

It does not work, since arange only takes integer arguments as size. I would suggest that this is implemented, otherwise it is a bit complicated to use arange (then reshape, then relabel).

### PR DESCRIPTION
**Pull request by jeffry1829**
Fixed #325

### CODE REVIEW

**Comment by kaihsin** [comment]
What is going on with the test No. 520 (BkUT_Norm)

**Comment by hunghaoti** [comment]
> What is going on with the test No. 520 (BkUT_Norm)

Hi @kaihsin , did you mean this modification:
df12df8
Because for original code user can use put_block to put the Tensor into the UniTensor even there are not same dType. After the commit 
https://github.com/Cytnx-dev/Cytnx/commit/58897042447f4c6ab4d18cad8a71b66c118f13bf#L123
this function will check whether the dType of the input tensor same as UniTensor, see
https://github.com/Cytnx-dev/Cytnx/blob/58897042447f4c6ab4d18cad8a71b66c118f13bf/include/UniTensor.hpp#L1430-L1433
So the original test code will throw the error at line 340, see 
https://github.com/Cytnx-dev/Cytnx/blob/276507974c4cc974291a7318ac906d0825dcdf77/tests/linalg_test/linalg_test.cpp#L330-L345

**Comment by kaihsin** [comment]
Why not just try to cast it, I don't know what is going on

File: include/Generator.hpp [hunk 1]
```diff
@@ -132,11 +142,56 @@ namespace cytnx {
...
+    @param dtype the dtype of the Tensor. It can be any type defined in \link cytnx::Type
+    cytnx::Type \endlink
+    @param device the device that the Tensor is put on. It can be any device defined in \link
+    cytnx::Device cytnx::Device \endlink
+
+    @return
+        [Tensor]
+
+    */
+  Tensor arange(const std::vector<cytnx_uint64> &shape, const cytnx_double &start,
```
Comment (kaihsin) [inline include/Generator.hpp:162]: Dont do this

File: include/Generator.hpp [hunk 2]
```diff
@@ -132,11 +142,56 @@ namespace cytnx {
...
+     evenly spaced. Then reshape to the shape specified by the input shape.
+      @param start the start value of the range
+      @param end the end value of the range
+      @param Nelem the number of elements
+      @param endpoint if true, the end value is included in the range.
+      @param dtype the dtype of the Tensor. It can be any type defined in \link cytnx::Type \endlink
+      @param device the device that the Tensor is put on. It can be any device defined in \link
+      cytnx::Device \endlink
+    */
+  Tensor linspace(const std::vector<cytnx_uint64> &shape, const cytnx_double &start,
```
Comment (kaihsin) [inline include/Generator.hpp:192]: remove

File: include/UniTensor.hpp [hunk 3]
```diff
@@ -4165,6 +4165,26 @@ namespace cytnx {
...
+    @param[in] Nelem the number of elements.
+    @param[in] in_labels the labels of the UniTensor.
+    @param[in] name the name of the UniTensor.
+    @return
+        [UniTensor]
+    @see arange(const std::vector<cytnx_uint64> &shape, const cytnx_double &start, const
+    cytnx_double &end, const cytnx_double &step, const unsigned int &dtype, const int &device) \n
+    @see arange(const cytnx_int64 &Nelem)
+    */
+    static UniTensor arange(const std::vector<cytnx_uint64> &shape,
```
Comment (kaihsin) [inline include/UniTensor.hpp:4182]: same as above

File: include/UniTensor.hpp [hunk 4]
```diff
@@ -4189,6 +4209,32 @@ namespace cytnx {
...
+    @param[in] in_labels the labels of the UniTensor.
+    @param[in] dtype the data type of the UniTensor, see cytnx::Type for more information.
+    @param[in] device the device type of the UniTensor, see cytnx::Device for more information.
+    @param[in] name the name of the UniTensor.
+    @return
+        [UniTensor]
+    @see arange(const std::vector<cytnx_uint64> &shape, const cytnx_double &start, const
+    cytnx_double &end, const cytnx_double &step, const unsigned int &dtype, const int &device)
+    */
+    static UniTensor arange(const std::vector<cytnx_uint64> &shape, const cytnx_double &start,
```
Comment (kaihsin) [inline include/UniTensor.hpp:4229]: same as above

File: include/UniTensor.hpp [hunk 5]
```diff
@@ -4217,6 +4263,35 @@ namespace cytnx {
...
+    @param[in] in_labels the labels of the UniTensor.
+    @param[in] dtype the data type of the UniTensor, see cytnx::Type for more information.
+    @param[in] device the device type of the UniTensor, see cytnx::Device for more information.
+    @param[in] name the name of the UniTensor.
+    @return
+        [UniTensor]
+    @see linspace(const cytnx_double &start, const cytnx_double &end, const cytnx_uint64 &Nelem,
+    const bool &endpoint, const unsigned int &dtype, const int &device)
+    */
+    static UniTensor linspace(const std::vector<cytnx_uint64> &shape, const cytnx_double &start,
```
Comment (kaihsin) [inline include/UniTensor.hpp:4285]: same above

File: pybind/generator_py.cpp
Comment (kaihsin) [inline pybind/generator_py.cpp:1]: same adjustment as c++ side

File: pybind/unitensor_py.cpp
Comment (kaihsin) [inline pybind/unitensor_py.cpp:1]: same adjustment as c++ side

File: src/Generator.cpp
Comment (kaihsin) [inline src/Generator.cpp:1]: please remove all the things with arange with shape as list

File: tests/DenseUniTensor_test.cpp
Comment (kaihsin) [inline tests/DenseUniTensor_test.cpp:1]: testing case accrodingly

**Review by kaihsin — CHANGES_REQUESTED** [review]

**Comment by kaihsin** [comment]
@hunghaoti , the user can always put unitensor with different type, you just need to cast it to the proper type. I think you code in C++ and Python enough to know it should be intrinsic casting for UX

**Comment by jeffry1829** [comment]
@kaihsin You mean we only keep `arange(list)`, and remove all other added function?

File: include/UniTensor.hpp (hunk 3 above)
Comment (jeffry1829) [reply include/UniTensor.hpp:4182]: I think this should be kept?
We should remove arange(shape, start,end,...) ?

Comment (kaihsin) [reply include/UniTensor.hpp:4182]: you are combining arange and reshape at the same API call. numpy does not have this kind of APIs, and its inconsistent

**Pull request closed without merging**